Tribler/Core/Upgrade/db72_to_pony.py

```diff
@@ -75,56 +76,53 @@ def initialize(self, mds):
             self._logger.info("No personal channel found")
 
     def get_old_channels(self):
-        connection = sqlite3.connect(self.tribler_db)
-        cursor = connection.cursor()
-
-        channels = []
-        for id_, name, dispersy_cid, modified, nr_torrents, nr_favorite, nr_spam in cursor.execute(
-                self.select_channels_sql):
-            if nr_torrents and nr_torrents > 0:
-                channels.append({"id_": infohash_to_id(dispersy_cid),
-                                 # converting this to str is a workaround for python 2.7 'writable buffers not hashable'
-                                 # problem with Pony
-                                 "infohash": bytes(dispersy_cid),
-                                 "title": name or '',
-                                 "public_key": b"",
-                                 "timestamp": final_timestamp(),
-                                 "origin_id": 0,
-                                 "size": 0,
-                                 "subscribed": False,
-                                 "status": LEGACY_ENTRY,
-                                 "votes": -1,
-                                 "num_entries": int(nr_torrents or 0)})
-        connection.close()
+        with contextlib.closing(sqlite3.connect(self.tribler_db)) as connection, connection:
```

Comment:
Why is there `as connection, connection`? What kind of idiom is this? I don't get it.

Reply:
What happens is that using a db connection with a bare "with" statement doesn't close it -- instead it commits whatever transaction it had in the block (and keeps the connection open). So, to close it, `contextlib.closing` is used. However, as the docs states (can be seen in the link I posted above), closing doesn't commit the transactions. Basically, two "with" blocks are needed to handle errors:
```python
# This closes the connection
with contextlib.closing(sqlite3.connect(self.tribler_db)) connection
    # This commits the transactions
    with connection:
        # Code
```
However, python allows to do this instead of nesting with statements:
```python
with f() as a, g() as b:
    # Code
```
And it closes them in the reverse order to their initial setup. Basically, it's syntax sugar to nested "with" statements.
So `with contextlib.closing(sqlite3.connect(self.tribler_db)) as connection, connection:` is equivalent to nesting them.

While going back to closing them manually is possible, it has a danger of a connection leaking when an exception happens since it won't be closed. This way allows for more robust error handling.

Reply:
Ok, I see now. This is a corner case of syntactic sugar usage that just happen to look weird.

Reply:
Yes, exactly